src/sqs.ts

```diff
@@ -81,19 +91,31 @@ export class SQSMessageHandler<Message, Context> {
 
       // 2. Process all the records.
       context.logger.info({ event }, 'Processing SQS topic message');
-      for (const record of event.Records) {
-        const messageLogger = context.logger.child({
-          messageId: record.messageId,
-        });
-
-        const parsedMessage = this.config.parseMessage(record.body);
-
-        for (const action of this.messageActions) {
-          await action({ ...context, logger: messageLogger }, parsedMessage);
-        }
 
-        messageLogger.info('Successfully processed message');
-      }
+      await processWithOrdering(
+        {
+          items: event.Records,
+          // If there is not a MessageGroupId, then we can don't care about
+          // the ordering for the event. We can just generate a UUID for the
+          // ordering key.
+          orderBy: (record) => record.attributes.MessageGroupId ?? uuid(),
+          concurrency: this.config.concurrency ?? 5,
+          stopOnError: false,
+        },
+        async (record) => {
+          const messageLogger = context.logger.child({
+            messageId: record.messageId,
+          });
+
+          const parsedMessage = this.config.parseMessage(record.body);
+
+          for (const action of this.messageActions) {
+            await action({ ...context, logger: messageLogger }, parsedMessage);
```

Comment:
What happens when a single message fails to be processed? Does it just get dropped if `stopOnError` is false? In the case that `stopOnError` is `true`, the whole message batch would be reprocessed, correct? In that case we're assuming that the `action` is idempotent and we leave it up to consumers to make sure of that?

Reply:
This is true for both SQS and DynamoDB streams, except streams can be divided on errors.

Reply:
Good question. If a single message fails, then it will be thrown once all other messages have been processed.

So, in the case of a failure, the entire batch will be reprocessed. This matches the current behavior of `SQSMessageHandler`.

The improvement here that it will attempt to process other messages even in the case of a failure.

See the docs for details: https://github.com/sindresorhus/p-map#stoponerror

Reply:
In SQS, it is possible to signal a "partial failure" and only retry the failed messages (see [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#services-sqs-batchfailurereporting).

But, `SQSMessageHandler` does support this behavior yet.

Reply:
That's awesome! I'm going to have to add that now...